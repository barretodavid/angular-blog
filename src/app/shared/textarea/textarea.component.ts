import { Component, ChangeDetectionStrategy, Self } from '@angular/core';
import { NgControl } from '@angular/forms';
import { BaseInputClass } from '../base-input/base-input.class';

@Component({
  selector: 'rg-textarea',
  template: `
    <mat-form-field appearance="standard" class="w-100">
      <mat-label>{{ label }}</mat-label>
      <textarea matInput [formControl]="controlDir.control"></textarea>
      <mat-error *ngIf="controlDir.control.hasError('required')">
        This field is required
      </mat-error>
    </mat-form-field>
  `,
  changeDetection: ChangeDetectionStrategy.OnPush
})
export class TextareaComponent extends BaseInputClass {
  constructor(@Self() public controlDir: NgControl) {
    super(controlDir);
  }
}
