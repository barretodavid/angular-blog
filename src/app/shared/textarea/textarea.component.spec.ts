import { Component } from '@angular/core';
import { NoopAnimationsModule } from '@angular/platform-browser/animations';
import {
  ReactiveFormsModule,
  FormControl,
  NgControl,
  Validators,
} from '@angular/forms';
import { ComponentFixture, async, TestBed } from '@angular/core/testing';
import { CommonModule } from '@angular/common';
import { MatFormFieldModule, MatInputModule } from '@angular/material';

import { TextareaComponent } from './textarea.component';

const formControl = new FormControl('Test', Validators.required);

@Component({
  selector: 'rg-fake',
  template:
    '<rg-textarea [formControl]="content" label="Content"></rg-textarea>',
})
class FakeComponent {
  content = formControl;
}

describe('TextareaComponent', () => {
  let component: FakeComponent;
  let fixture: ComponentFixture<FakeComponent>;
  let textarea: HTMLTextAreaElement;

  beforeEach(async(() => {
    TestBed.configureTestingModule({
      imports: [
        NoopAnimationsModule,
        CommonModule,
        MatFormFieldModule,
        MatInputModule,
        ReactiveFormsModule,
      ],
      declarations: [FakeComponent, TextareaComponent],
    })
      .overrideComponent(TextareaComponent, {
        set: {
          providers: [
            {
              provide: NgControl,
              useValue: formControl,
            },
          ],
        },
      })
      .compileComponents();
  }));

  beforeEach(() => {
    fixture = TestBed.createComponent(FakeComponent);
    component = fixture.componentInstance;
    textarea = fixture.nativeElement.querySelector('textarea');
    fixture.detectChanges();
  });

  test('Should render', () => {
    expect(fixture).toMatchSnapshot();
  });

  describe('When clearing the textarea and losing focus', () => {
    test('Should render an error message', () => {
      component.content.patchValue('');
      fixture.detectChanges();
      textarea.dispatchEvent(new Event('blur'));
      fixture.detectChanges();
      expect(fixture).toMatchSnapshot();
    });
  });

  describe('When adding a value to the textarea and losing focus', () => {
    test('Should not render an error message', () => {
      component.content.patchValue('Some content');
      fixture.detectChanges();
      textarea.dispatchEvent(new Event('blur'));
      fixture.detectChanges();
      expect(fixture).toMatchSnapshot();
    });
  });
});
