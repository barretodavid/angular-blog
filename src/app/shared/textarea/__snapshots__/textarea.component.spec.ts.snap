// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`TextareaComponent Should render 1`] = `
<rg-fake
  content={[Function FormControl]}
>
  <rg-textarea
    class="ng-untouched ng-pristine ng-valid"
    label="Content"
    ng-reflect-form="[object Object]"
    ng-reflect-label="Content"
  >
    <mat-form-field
      appearance="standard"
      class="w-100 mat-form-field ng-tns-c0-0 mat-primary mat-form-field-type-mat-input mat-form-field-appearance-standard mat-form-field-can-float mat-form-field-should-float ng-untouched ng-pristine ng-valid _mat-animation-noopable"
      ng-reflect-appearance="standard"
    >
      <div
        class="mat-form-field-wrapper"
      >
        <div
          class="mat-form-field-flex"
        >
          
          
          <div
            class="mat-form-field-infix"
          >
            <textarea
              aria-invalid="false"
              aria-required="false"
              class="mat-input-element mat-form-field-autofill-control cdk-text-field-autofill-monitored ng-untouched ng-pristine ng-valid"
              id="mat-input-0"
              matinput=""
              ng-reflect-form="[object Object]"
            />
            <span
              class="mat-form-field-label-wrapper"
            >
              
              <label
                aria-owns="mat-input-0"
                class="mat-form-field-label ng-tns-c0-0 ng-star-inserted"
                for="mat-input-0"
                id="mat-form-field-label-1"
                ng-reflect-disabled="true"
                ng-reflect-ng-switch="true"
              >
                
                
                <mat-label
                  class="ng-star-inserted"
                >
                  Content
                </mat-label>
                
                
              </label>
            </span>
          </div>
          
        </div>
        
        <div
          class="mat-form-field-underline ng-tns-c0-0 ng-star-inserted"
        >
          <span
            class="mat-form-field-ripple"
          />
        </div>
        <div
          class="mat-form-field-subscript-wrapper"
          ng-reflect-ng-switch="hint"
        >
          
          
          <div
            class="mat-form-field-hint-wrapper ng-tns-c0-0 ng-trigger ng-trigger-transitionMessages ng-star-inserted"
            style="opacity:1;transform:translateY(0%);0:opacity;1:transform;"
          >
            
            <div
              class="mat-form-field-hint-spacer"
            />
          </div>
        </div>
      </div>
    </mat-form-field>
  </rg-textarea>
</rg-fake>
`;

exports[`TextareaComponent When adding a value to the textarea and losing focus Should not render an error message 1`] = `
<rg-fake
  content={[Function FormControl]}
>
  <rg-textarea
    class="ng-touched ng-pristine ng-valid"
    label="Content"
    ng-reflect-form="[object Object]"
    ng-reflect-label="Content"
  >
    <mat-form-field
      appearance="standard"
      class="w-100 mat-form-field ng-tns-c2-0 mat-primary mat-form-field-type-mat-input mat-form-field-appearance-standard mat-form-field-can-float ng-touched ng-pristine _mat-animation-noopable mat-form-field-should-float ng-valid"
      ng-reflect-appearance="standard"
    >
      <div
        class="mat-form-field-wrapper"
      >
        <div
          class="mat-form-field-flex"
        >
          
          
          <div
            class="mat-form-field-infix"
          >
            <textarea
              aria-invalid="false"
              aria-required="false"
              class="mat-input-element mat-form-field-autofill-control cdk-text-field-autofill-monitored ng-touched ng-pristine ng-valid"
              id="mat-input-2"
              matinput=""
              ng-reflect-form="[object Object]"
            />
            <span
              class="mat-form-field-label-wrapper"
            >
              
              <label
                aria-owns="mat-input-2"
                class="mat-form-field-label ng-tns-c2-0 ng-star-inserted"
                for="mat-input-2"
                id="mat-form-field-label-5"
                ng-reflect-disabled="true"
                ng-reflect-ng-switch="true"
              >
                
                
                <mat-label
                  class="ng-star-inserted"
                >
                  Content
                </mat-label>
                
                
              </label>
            </span>
          </div>
          
        </div>
        
        <div
          class="mat-form-field-underline ng-tns-c2-0 ng-star-inserted"
        >
          <span
            class="mat-form-field-ripple"
          />
        </div>
        <div
          class="mat-form-field-subscript-wrapper"
          ng-reflect-ng-switch="hint"
        >
          
          
          <div
            class="mat-form-field-hint-wrapper ng-tns-c2-0 ng-trigger ng-trigger-transitionMessages ng-star-inserted"
            style="opacity:1;transform:translateY(0%);0:opacity;1:transform;"
          >
            
            <div
              class="mat-form-field-hint-spacer"
            />
          </div>
        </div>
      </div>
    </mat-form-field>
  </rg-textarea>
</rg-fake>
`;

exports[`TextareaComponent When clearing the textarea and losing focus Should render an error message 1`] = `
<rg-fake
  content={[Function FormControl]}
>
  <rg-textarea
    class="ng-pristine ng-invalid ng-touched"
    label="Content"
    ng-reflect-form="[object Object]"
    ng-reflect-label="Content"
  >
    <mat-form-field
      appearance="standard"
      class="w-100 mat-form-field ng-tns-c1-0 mat-primary mat-form-field-type-mat-input mat-form-field-appearance-standard mat-form-field-can-float ng-pristine _mat-animation-noopable mat-form-field-hide-placeholder ng-invalid mat-form-field-invalid ng-touched"
      ng-reflect-appearance="standard"
    >
      <div
        class="mat-form-field-wrapper"
      >
        <div
          class="mat-form-field-flex"
        >
          
          
          <div
            class="mat-form-field-infix"
          >
            <textarea
              aria-describedby="mat-error-0"
              aria-invalid="true"
              aria-required="false"
              class="mat-input-element mat-form-field-autofill-control cdk-text-field-autofill-monitored ng-pristine ng-invalid ng-touched"
              id="mat-input-1"
              matinput=""
              ng-reflect-form="[object Object]"
            />
            <span
              class="mat-form-field-label-wrapper"
            >
              
              <label
                aria-owns="mat-input-1"
                class="mat-form-field-label ng-tns-c1-0 ng-star-inserted mat-empty mat-form-field-empty"
                for="mat-input-1"
                id="mat-form-field-label-3"
                ng-reflect-disabled="true"
                ng-reflect-ng-switch="true"
              >
                
                
                <mat-label
                  class="ng-star-inserted"
                >
                  Content
                </mat-label>
                
                
              </label>
            </span>
          </div>
          
        </div>
        
        <div
          class="mat-form-field-underline ng-tns-c1-0 ng-star-inserted"
        >
          <span
            class="mat-form-field-ripple"
          />
        </div>
        <div
          class="mat-form-field-subscript-wrapper"
          ng-reflect-ng-switch="error"
        >
          
          <div
            class="ng-tns-c1-0 ng-trigger ng-trigger-transitionMessages ng-star-inserted ng-animating"
            style=""
          >
            
            <mat-error
              class="mat-error ng-star-inserted"
              id="mat-error-0"
              role="alert"
            >
               This field is required 
            </mat-error>
          </div>
          
        </div>
      </div>
    </mat-form-field>
  </rg-textarea>
</rg-fake>
`;
